### DISCUSSION TITLE
Can you use `useState` at the top level of an `.mdx` file?

### DISCUSSION THREAD

**Discussion by oconnoob** [facebook/docusaurus#9056] [Q&A] [answered]
Hi there,

I'm wondering if it is possible to use state at the top level of an `.mdx` file? I'm looking to achieve something like this:

```javascript
// myfile.mdx
import { useState } from "react";

export const [myState, setMyState] = useState(false)

// maybe have to use a callback instead?
<button onClick={() => setMyState(true)}>Set true</button>

# Section header

Here is some text that is normal markdown

<>{myState 
? "here is a paragraph that should be shown when `myState` is `true`" 
: ? "here is a paragraph that should be shown when `myState` is `false`"}
</>
```

I effectively want to maintain the usefulness of markdown (easy to parse and write) but be able to conditionally show different markdown depending on state. 

Any insight is appreciated! I've been scratching my head on this for a while

**Comment by slorber** [comment, marked as answer]
No, it is not possible as far as I know, although I wish it was (cc @wooorm) but I don't think it's planned?

If you try to use the [MDX playground](https://mdxjs.com/playground/) (unfortunately does not really make it possible to use imports here), a top-level useState will compile to something like this:

```tsx
/*@jsxRuntime automatic @jsxImportSource react*/
const {Fragment: _Fragment, jsx: _jsx} = arguments[0];
const [myState, setMyState] = React.useState(false);
function _createMdxContent(props) {
  return _jsx(_Fragment, {});
}
function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? _jsx(MDXLayout, Object.assign({}, props, {
    children: _jsx(_createMdxContent, props)
  })) : _createMdxContent(props);
}
return {
  myState,
  setMyState,
  default: MDXContent
};
```

This should work better but you'd have to figure out how to interleave MD / JSX

```tsx
import { useState } from "react";

export function MyComponent() {
  const [myState, setMyState] = useState(false);
  return <>{myState}</>
}

<MyComponent/>
```

I'd recommend keeping complex/stateful logic outside of MDX. Eventually, you can use a state provider around the MDX component that provides the stateful logic. This is what Docusaurus does for some features like tabs for example.

This answer is for MDX v2 (upcoming in Docusaurus v3) but I doubt it would be much different for MDX v1 (Docusaurus v2)

**Reply by oconnoob** [reply]
Thank you for this succinct answer!

**Comment by wooorm** [comment]
* Q: I want to have MDX in strings that is evaluated
  A: This isn’t possible in JS either: `console.log("alert(1)")` doesn’t show an alert. It shows a string. Evaluating code at runtime is a bad bad idea. It’s slow. It’s unsafe. There are good alternatives.

The alternatives are:

`message-positive.mdx`

```mdx
here is a paragraph that should be shown when `myState` is `true`
```

`message-negative.mdx`

```mdx
here is a paragraph that should be shown when `myState` is `false`
```

`example.jsx`

```jsx
<>{myState ? <MessagePositive /> : <MessageNegative />}</>
```

---

You can additionally use an `If` component, something that does:

```jsx
<If condition={myState}>
  here is a paragraph that should be shown when `myState` is `true`
</If>
<If condition={!myState}>
  here is a paragraph that should be shown when `myState` is `false`
</If>
```

**Reply by slorber** [reply]
Thanks.

The question was more related to where the state comes from, ie declaring `useState(false)` at the top/root of the MDX file instead of inside an inlined component

**Reply by wooorm** [reply]
Yeah I felt like you already responded well on that :)

You know, https://github.com/mdx-js/mdx/issues/701#issuecomment-947030041 is 90% this, you’d have to figure out if it’s about *expressions* or *statements* tho. Alternatively one could use a frontmatter syntax that includes only JS statements at the top and take the same approach?
It’s an interesting idea, open to discussing it more.

**Reply by oconnoob** [reply]
Thanks for your comment @wooorm
